triedb/pathdb/history_inspect.go

```diff
@@ -0,0 +1,149 @@
+// Copyright 2024 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/
+
+package pathdb
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/rawdb"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+// HistoryStats wraps the history inspection statistics.
+type HistoryStats struct {
+	Start   uint64   // Block number of the first queried history
+	End     uint64   // Block number of the last queried history
+	Blocks  []uint64 // Blocks refers to the list of block numbers in which the state is mutated
+	Origins [][]byte // Origins refers to the original value of the state before its mutation
+}
+
+// sanitizeRange limits the given range to fit within the local history store.
+func sanitizeRange(start, end uint64, freezer *rawdb.ResettableFreezer) (uint64, uint64, error) {
+	// Load the id of the first history object in local store.
+	tail, err := freezer.Tail()
+	if err != nil {
+		return 0, 0, err
+	}
+	first := tail + 1
+	if start != 0 && start > first {
+		first = start
+	}
+	// Load the id of the last history object in local store.
+	head, err := freezer.Ancients()
+	if err != nil {
+		return 0, 0, err
+	}
+	last := head - 1
+	if end != 0 && end < last {
+		last = end
+	}
+	// Make sure the range is valid
+	if first >= last {
+		return 0, 0, fmt.Errorf("range is invalid, first: %d, last: %d", first, last)
+	}
+	return first, last, nil
+}
+
+func inspectHistory(freezer *rawdb.ResettableFreezer, start, end uint64, onHistory func(*history, *HistoryStats)) (*HistoryStats, error) {
+	var (
+		stats  = &HistoryStats{}
+		init   = time.Now()
+		logged = time.Now()
+	)
+	start, end, err := sanitizeRange(start, end, freezer)
+	if err != nil {
+		return nil, err
+	}
+	for id := start; id <= end; id += 1 {
+		h, err := readHistory(freezer, id)
```

Comment:
For `accountHistory`, the `readHistory` would could skip this: 
```
		storageData    = rawdb.ReadStateStorageHistory(freezer, id)
		storageIndexes = rawdb.ReadStateStorageIndex(freezer, id)
```
It should speed it up considerably. 
For `storageHistory`, the `decode` part of the `dec.decode(accountData, storageData, accountIndexes, storageIndexes)` could omit everything but the storages for the account we're interested in. 

Maybe not worth optimizing? Up to you, just a thought

Reply:
Yep for sure we can optimize it. I will just add a todo, we can do it in a following PR.